frame/contracts/src/exec.rs

```diff
@@ -853,14 +853,19 @@ where
 			}
 			if let CachedContract::Cached(contract) = frame.contract_info {
 				// optimization: Predecessor is the same contract.
-				// We can just merge into the predecesor without a storage write.
+				// We can just copy the contract into the predecesor without a storage write.
+				// This is possible between there is no other contract in-between that could
+				// trigger a rollback.
 				if prev.account_id == *account_id {
 					prev.contract_info = CachedContract::Cached(contract);
 					return;
 				}
 
-				// Predecessor is a different contract: Write to storage and invalidate all
-				// references to it.
+				// Predecessor is a different contract: We persist the info and invalidate the first
+				// stale cache we find. This triggers a reload from storage on next use. We skip(1)
+				// because that case is already handled by the optimization above. Only the first
+				// cache needs to be invalidated because that one will invalidate the next cache
+				// once persisted.
```

Comment:
once persisted? I guess it's more fair to say once its frame is popped off or something like that